File: 393-utf-8-validation/utf-8-validation.py
class Solution:
    def validUtf8(self, data):
        remaining = 0

        for byte in data:

            if remaining == 0:

                mask = 128
                count = 0

                while byte & mask:
                    count += 1
                    mask >>= 1

                if count == 0:
                    continue

                if count == 1 or count > 4:
                    return False

                remaining = count - 1

            else:

                if byte >> 6 != 0b10:
                    return False

                remaining -= 1

        return remaining == 0

File: 393-utf-8-validation/utf-8-validation.java
class Solution {
    public boolean validUtf8(int[] data) {

        int remaining = 0;

        for (int b : data) {

            if (remaining == 0) {

                int mask = 128;
                int count = 0;

                while ((b & mask) != 0) {
                    count++;
                    mask >>= 1;
                }

                if (count == 0) {
                    continue;
                }

                if (count == 1 || count > 4) {
                    return false;
                }

                remaining = count - 1;
            }

            else {

                if ((b >> 6) != 2) {
                    return false;
                }

                remaining--;
            }
        }

        return remaining == 0;
    }
}